processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -556,9 +541,9 @@ public DimensionHolder getDimValues()
     return dimValues;
   }
 
-  public List<String> getDimensions()
+  public String[] getDimensions()
   {
-    return dimensions;
+    return dimensions.toArray(new String[dimensions.size()]);
```

Comment:
There are quite a few places in the code that call `getDimensions()` within loops or that make the assumptions it's a cheap operation. I am worried this will cause both performance problems, as well as increased garbage during merge and persist. I think this needs a bit more work to guarantee that this does not hurt query performance or increase heap / gc pressure during merge and persist.